src/Features/LanguageServer/Protocol/Extensions/ProtocolConversions.cs

```diff
@@ -151,35 +156,146 @@ static async Task<char> GetInsertionCharacterAsync(Document document, int positi
 
-            return new Uri(filePath, UriKind.Absolute);
+            if (PathUtilities.IsUnixLikePlatform)
+            {
+                // Unix path: first part is empty, all parts should be escaped
+                return "file://" + string.Join("/", parts.Select(part => EscapeUriPart(part.AsSpan())));
+            }
+
+            if (parts is ["", "", var serverName, ..])
+            {
+                // UNC path: first non-empty part is server name and shouldn't be escaped
+                return "file://" + serverName + "/" + string.Join("/", parts.Skip(3).Select(part => EscapeUriPart(part.AsSpan())));
+            }
+
+            // Drive-rooted path: first part is "C:" and shouldn't be escaped
+            return "file:///" + parts[0] + "/" + string.Join("/", parts.Skip(1).Select(part => EscapeUriPart(part.AsSpan())));
         }
 
-        public static Uri GetUriFromPartialFilePath(string? filePath)
+        public static Uri CreateUriFromSourceGeneratedFilePath(string filePath)
         {
-            if (filePath is null)
-                throw new ArgumentNullException(nameof(filePath));
+            Debug.Assert(!PathUtilities.IsAbsolute(filePath));
+
+            // Fast path for common cases:
+            if (IsAscii(filePath))
+            {
+                return new Uri(s_sourceGeneratedDocumentBaseUri, filePath);
+            }
+
+            // Workaround for https://github.com/dotnet/runtime/issues/89538:
 
-            return new Uri(SourceGeneratedDocumentBaseUri, filePath);
+            var parts = filePath.Split(s_dirSeparators);
+            var url = SourceGeneratedDocumentBaseUri + string.Join("/", parts.Select(Uri.EscapeDataString));
+
+            return new Uri(url, UriKind.Absolute);
         }
 
-        public static Uri? TryGetUriFromFilePath(string? filePath, RequestContext? context = null)
+        private static bool IsAscii(char c)
+            => (uint)c <= '\x007f';
+
+        private static bool IsAscii(string filePath)
         {
-            if (Uri.TryCreate(filePath, UriKind.Absolute, out var uri))
-                return uri;
+            for (var i = 0; i < filePath.Length; i++)
+            {
+                if (!IsAscii(filePath[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static BitVector CreateUriUnescapedBitVector()
+        {
+            // Indexed by ASCII character [0..80)
+            var map = BitVector.Create(0x80);
+
+            foreach (var c in "!$&'()*+,-./0123456789:;=?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[]_abcdefghijklmnopqrstuvwxyz~")
+            {
+                map[c] = true;
+            }
+
+            return map;
+        }
+
+        // Based on EscapeString in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.Uri/src/System/UriHelper.cs
+        private static string EscapeUriPart(ReadOnlySpan<char> stringToEscape)
```

Comment:
Just kill me.

Reply:
So I shouldn't ask for this to be made public API so I can call it from Razor? 😁

Reply:
We can expose it via External Access